cheevos/cheevos.c

```diff
@@ -1388,27 +1359,92 @@ static int cheevos_test_cond_set(const cheevos_condset_t *condset,
       cheevos_locals.add_buffer = 0;
       cheevos_locals.add_hits   = 0;
 
-      /* Sequential or non-sequential? */
-      set_valid &= cond_valid;
+      if (cond->type == CHEEVOS_COND_TYPE_PAUSE_IF)
+      {
+         /* as soon as we find a PauseIf that evaluates to true,
+          * stop processing the rest of the group. */
+         if (cond_valid)
+            return 1;
+
+         /* if we make it to the end of the function, make sure we are
+          * indicating nothing matched. if we do find a later PauseIf match,
+          * it'll automatically return true via the previous condition. */
+         set_valid = 0;
+
+         if (cond->req_hits == 0)
+         {
+            /* PauseIf didn't evaluate true, and doesn't have a HitCount,
+             * reset the HitCount to indicate the condition didn't match. */
+            if (cond->curr_hits != 0)
+            {
+               cond->curr_hits = 0;
+               *dirty_conds = 1;
+            }
+         }
+         else
+         {
+            /* PauseIf has a HitCount that hasn't been met, ignore it for now. */
+         }
+      }
+      else if (cond->type == CHEEVOS_COND_TYPE_RESET_IF)
+      {
+         if (cond_valid)
+         {
+            *reset_conds = 1; /* Resets all hits found so far */
+            set_valid    = 0; /* Cannot be valid if we've hit a reset condition. */
+         }
+      }
+      else /* Sequential or non-sequential? */
+         set_valid &= cond_valid;
    }
 
-   /* Now, ONLY read reset conditions! */
-   for (cond = condset->conds; cond < end; cond++)
-   {
-      if (cond->type != CHEEVOS_COND_TYPE_RESET_IF)
-         continue;
+   return set_valid;
+}
 
-      cond_valid = cheevos_test_condition(cond);
+static int cheevos_test_pause_cond_set(const cheevos_condset_t *condset,
+      int *dirty_conds, int *reset_conds)
+{
+   if (!condset)
+      return 1; /* important: empty group must evaluate true */
```

Comment:
should you also check for `condset->count`?

Reply:
if you mean that I should check `condset->count == 0` then no.

if the condition set is empty, then `condset == NULL`, which is evaluated to false, but with the `!` operator it is converted to true.

Reply:
`condset` is a pointer to a struct that has a `count` member, the existing check just makes sure the pointer is not null. my question is whether you also need to validate the `count` is non-zero. i.e. what does this function do if passed an empty group instead of a non-existent one.

Reply:
I'm repeating the current implementation:
https://github.com/libretro/RetroArch/blob/master/cheevos/cheevos.c#L1300

I didn't actually checked, but as it's working fine currently, then I guess that empty groups don't reach this part of the logic.

But, OK, I will double check.